### PR TITLE
implement cooldown

### PR DESCRIPTION
**Pull request by joe-p**

### CODE REVIEW

**Comment by kylebeee** [comment]
I love this and i think its a great compromise for the concerns raised by D13. I say merge it

File: contracts/abstracted_account.algo.ts [hunk 1]
```diff
@@ -124,19 +126,31 @@ export class AbstractedAccount extends Contract {
...
    */
   arc58_rekeyToPlugin(plugin: AppID): void {
-    const globalKey: PluginsKey = { application: plugin, allowedCaller: globals.zeroAddress };
-
-    // If this plugin is not approved globally, then it must be approved for this address
-    if (!this.plugins(globalKey).exists || this.plugins(globalKey).value < globals.latestTimestamp) {
-      const key: PluginsKey = { application: plugin, allowedCaller: this.txn.sender };
-      assert(this.plugins(key).exists && this.plugins(key).value > globals.latestTimestamp);
-    }
+    assert(this.pluginCallAllowed(plugin, this.txn.sender), 'This sender is not allowed to trigger this plugin');
```
Comment (kylebeee) [inline contracts/abstracted_account.algo.ts:153]: We need lastCalled to be updated for cooldowns to work properly

```typescript
    // update the round the plugin was last called
    this.updatePluginLastCalled(plugin, this.txn.sender);
```

```typescript
  private updatePluginLastCalled(app: AppID, caller: Address): void {
    const key: PluginsKey = { application: app, allowedCaller: caller };
    this.plugins(key).value.lastCalled = globals.round;
  }
```

Comment (joe-p) [reply contracts/abstracted_account.algo.ts:153]: Good catch. The more I think about it the more having the address as the key is getting annoying. If someone wants  to have the same functionality for multiple addresses then they can just create multiple of the same app. I'm going to change it so the address is not in the key and then fix this after

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts:153]: ahh i think we need to keep it in there incase of delegations made to multiple different entities. Forcing new apps with the same exact code to exist i think might create more confusion & limit functionality

Comment (joe-p) [reply contracts/abstracted_account.algo.ts:153]: How so? Hashes of programs are no more or less recognizable than huge uint64s. In both scenarios the admin would need to add plugins separately so really the only difference is how we identify the plugin to the user, which we could do via the program. Side benefit is plugins can be identified across networks.

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts:153]: off the top of my head, programs the address you're delegating to doesn't have control of. Apps aren't just their code, they're also the assets the program has in its account. You can mint another AMM contract but that doesn't mean it would be a helpful thing to do.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts:153]: I envision it as the AMM has one app ID and the plugin for that AMM is a completely separately app (in this case with multiple deployments). If you mix core app logic with plugin logic plugins become much harder to audit.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts]: I also think it's what most plugins are going to be, at least short/mid term. It's unrealistic to expect every defi protocol to update their existing apps, but it's much more reasonable to create plugins for them. 

Take a look at [774cf1c](https://github.com/joe-p/arc58/pull/9/commits/774cf1c2e3291d64743ed3c20df73ac874b1c219). It simplifies the code on chain and for the caller since less box refs are needed which is nice.

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts]: > If you mix core app logic with plugin logic plugins become much harder to audit

You can argue that this version of plugins just means you now need to audit the plugin + the app where the core functionality lives. It doesn't make auditing any easier, potentially harder.

I also think that going that route undermines some of the security properties plugins should have. The plugin being immutable is meaningless if all it does is call another app that can be updated at anytime.

Additionally, i expect some security / safety to be lost if we have many app plugins that just call xyz contract, this would complicate the process of ensuring malicious plugins aren't getting added to peoples accounts imo.

I agree that a likely adoption curve for this is existing dapps just making a plugin that calls their other code. Although i hope to discourage it for the reasons stated. I think ideally we allow a graceful fallback to what we currently have & the plugin architecture can be adopted when builders are ready to support a version of their app with the constraints of a wallet plugin.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts]: > You can argue that this version of plugins just means you now need to audit the plugin + the app where the core functionality lives. It doesn't make auditing any easier, potentially harder.

You're auditing for two completely different things though. The plugin is your entire account. The app only controls whatever you send it. The burden of the audit for the app does not change with external plugins, but if you add plugins into the app it becomes more complex to audit them because it's a monolithic code base. Having a monolithic codebase also undermines the value of static analysis for plugins because complex apps might have a lot of inner transactions. 

I also think it's very likely that there will be many different plugins for one app. For example, defi apps could have multiple different aggregators, auto-compounders, limit orders, etc. from a bunch of different protocols that implement it differently. It's unreasonable to expect the DeFi protocol to bake that all into the core protocol app(s) themselves.

> Additionally, i expect some security / safety to be lost if we have many app plugins that just call xyz contract, this would complicate the process of ensuring malicious plugins aren't getting added to peoples accounts imo.

The security concerns of a plugin calling an app is no different than a user calling the app. If I trust an AMM enough to send it 100 ALGO how is allowing a plugin to send it 100 ALGO any different?

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts]: I think removing the address from the key on this basis is prioritizing separation of concerns over pragmatism. All the conversations we've had about how to ensure safety are implicitly about how to ensure bad things dont happen while the plugin has control of your account. If the plugin is immutable but all it does is call a mutable contract we dont gain anything & we complicate the process of ensuring safety.

I see where you're coming from but at the same time questions like: "should i have to go mint a new plugin if my user already has this plugin i want to call on their behalf added?" & "how do we approach sharing these plugins that all they do is proxy to another contract if i am to have to mint one in that circumstance" come to mind. Additionally, i feel strongly that keeping plugins as simple proxies wont be feasible and we're all but guaranteed to have a good mix of both.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts]: > I think removing the address from the key on this basis is prioritizing separation of concerns over pragmatism.

It definitely is prioritizing separation of concerns but I don't think it's any less pragmatic. 

> If the plugin is immutable but all it does is call a mutable contract we dont gain anything 

The value of this ARC for EOAs is being able to delegate control of an account so the address can send transactions without requiring a signature. If I trust an AMM to send it 100 ALGO why should I not trust a plugin that does the same thing? The plugin is just about not having to explicitly sign. Trusting the app is a completely separate issue. 

Trying to get users to not interact with mutable apps is an entirely separate issue from this ARC and there are lower-level ways of solving the question of "is this app mutable" that can be applied broadly to all transactions, ARC58, EOA, or any other context (ie. standardizing how to prevent `UpdateApplication`).

> & we complicate the process of ensuring safety.

I think encouraging apps to have monolithic code bases is compromising safety. If I trust a plugin I am trusting it with my entire account. If I trust an app I am only trusting it with what I send it. If we encourage developers to mingle these two together, the lines will get blurred.

> "should i have to go mint a new plugin if my user already has this plugin i want to call on their behalf added?" 

This is already a question that dApps have to answer with the current design except in the form of "do I need to call `arc58_addPlugin`?". In both scenarios the dApp needs to look at app state to determine if a new plugin is needed or not. 

> "how do we approach sharing these plugins that all they do is proxy to another contract if i am to have to mint one in that circumstance"

I'm not really convinced that deploying multiple instances of plugins is substantially more complex than having to add and use plugins to begin with. 

There is nothing that becomes impossible with this change, it's simply a different architecture. Yes it puts a little more onus on developers (if they want to co-mingle business and plugin logic), but keeping ARC58 itself simple is something that we should prioritize. 

In short, I am a big believer in the unix philosophy. The thing ARC58 does well is delegation. Concerns over app mutability is a separate problem.

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts]: I think we fundamentally disagree on the purpose of arc58 then. I see it as a paradigm shift for account UX and it feels like you see it as simply a means to have delegation and i think you're being dismissive of the complexity being added by removing the address in regards to the practical considerations of implementing support for these types of accounts.

Lets consider the scenario in which i want to allow 15 of my friends to opt me into assets. Now i need 15 different optin plugins? That doesn't seem like a practical approach to me. 

Im very much in favor of composability but mutability in the context of delegation is a much bigger risk.

If the appId is the key we get an easier to compose boxKey and thats really it.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts]: > I see it as a paradigm shift for account UX and it feels like you see it as simply a means to have delegation

ARC58 is a building block for a paradigm shift in account UX. It is the building block that is really good at handling account delegation. That is the fundamental problem for account abstraction in any context. 

> Im very much in favor of composability but mutability in the context of delegation is a much bigger risk.

I am not advocating for mutable plugins. I am saying we should encourage developers to separate their business logic from plugin logic.

> If the appId is the key we get an easier to compose boxKey and thats really it.

I think the simplicity in the ARC58 logic itself is also good. To be clear though, I don't really have a super strong opinion about the box key itself. This conversation kinda became more about the philosophy of how we should encourage developers to develop plugins in relation to their apps. 

I am open to reverting the box key, but I think we should have clear reasoning for why we're adding more complexity to the contract. 

I also just realized that a downside to this change is the MBR requirements for each app which might hurt scalability... so yeah I will revert back to address in key, but I think we should keep this discussion about development philosophy in mind as the ARC matures.

Comment (kylebeee) [reply contracts/abstracted_account.algo.ts:153]: > I am not advocating for mutable plugins. I am saying we should encourage developers to separate their business logic from plugin logic.

What I've been poorly attempting to communicate is that immutable plugins don't mean anything if they simply call other mutable apps. Since delegations will likely be 'set & forget' any mutable contracts in the chain of calls represents a high degree of risk. The only reason this piece became part of the conversation was because althought we're likely to see a massive amount of plugins get made there will be staple app IDs that i hope become recognizable and trusted vs tons of different IDs accomplishing the same thing & i felt that change was a step backwards in that regards.

> This conversation kinda became more about the philosophy of how we should encourage developers to develop plugins in relation to their apps.

It may not be the best place for it but i'm very glad its something we're considering. I know the positions im advocating for imply a higher barrier to adoption but I think we should rely fairly heavily on `arc58_rekeyTo` as a means for graceful fallback & I think we could improve upon this aspect. Possibly allow anyone to rekey back to the contract so plugins could still work.

Apologies if I've come across as overly combative here. I feel rather strongly that embedding the delegation address in the key was an elegant method of creating key uniqueness & preserving some costs.

Comment (joe-p) [reply contracts/abstracted_account.algo.ts:153]: > What I've been poorly attempting to communicate is that immutable plugins don't mean anything if they simply call other mutable apps

I don't think it's so black & white though. Potentially loosing some amount of a specific ASA to a DeFi contract is way different than losing literally everything in my account. There are good reasons for mutability and some of the top protocols on Algorand have used mutable contracts. Also I've heard that anything MiCA-compliant has to be mutable, but not sure how accurate that is.

>  Since delegations will likely be 'set & forget' any mutable contracts in the chain of calls represents a high degree of risk.

The timestamp feature is what will discourage that set & forget mindset that we commonly see on EVM. 

I fully understand your concern over mutability, I just think it's a problem outside the scope of this ARC. Plugins don't really change much. You can send assets to a contract without much thought and it's the same kind of "set & forget". In fact I'd argue it's even safer now because without ARC58 contracts might require 10k ALGO up front, but with ARC58 they can request 100 ALGO every 100 rounds for 10k rounds. If an exploit happens those that sent 10k lose it all, but those that used a plugin can revoke.
